src/librbd/api/Mirror.cc

```diff
@@ -134,6 +136,135 @@ struct C_ImageGetStatus : public C_ImageGetInfo {
   }
 };
 
+int map_instance_id_to_service_id(librados::IoCtx& io_ctx,
+                                  std::map<std::string, std::string> *ids_map) {
+  CephContext *cct = reinterpret_cast<CephContext *>(io_ctx.cct());
+
+  std::string cmd = "{\"prefix\": \"service status\"}";
+
+  bufferlist in_bl;
+  bufferlist out_bl;
+  int r = librados::Rados(io_ctx).mgr_command(cmd, in_bl, &out_bl, nullptr);
+  if (r < 0) {
+    lderr(cct) << "failed to get service status: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  bool json_valid = false;
+  json_spirit::mValue json_root;
+  if (json_spirit::read(out_bl.to_str(), json_root)) {
+    try {
+      auto& json_obj = json_root.get_obj();
+      if (json_obj.count("rbd-mirror")) {
+        auto &mirror_service = json_obj["rbd-mirror"].get_obj();
+        for (auto &it : mirror_service) {
+          auto &service_id = it.first;
+          auto &daemon = it.second.get_obj();
+          if (daemon.count("status") &&
+              daemon["status"].get_obj().count("json")) {
+            auto& status_json_str =
+                daemon["status"].get_obj()["json"].get_str();
+            json_spirit::mValue status_json_root;
+            if (json_spirit::read(status_json_str, status_json_root)) {
+              auto& status = status_json_root.get_obj();
+              for (auto &iter : status) {
```

Comment:
The dict key should be the pool id, so you can just skip right to the pool's ```json``` status if it exists.